### PR TITLE
Formally define `SlowTest` / `QuickTest`

### PR DESCRIPTION
**Pull request by JackPGreen**
PR-builder time is growing, and the `SlowTest` / `QuickTest` annotations are meant to address that - but without a clear definition of the distinction between the two, many tests are probably in the wrong place.

Following a [Slack discussion](https://hazelcast.slack.com/archives/C01JU7ZJYGP/p1697440607109029), we've tentatively agreed 60 seconds is the threshold.

From this, I've written a simple application to scrape Jenkins to gather metrics on the average test time taken in the PR builder per test, over the last N builds:
```
import com.fasterxml.jackson.core.JsonParseException;
import com.fasterxml.jackson.databind.*;
import com.google.common.collect.*;
import com.hazelcast.test.annotation.*;

import java.net.URI;
import java.net.http.*;
import java.net.http.HttpResponse.BodyHandlers;
import java.text.NumberFormat;
import java.util.*;
import java.util.Map.Entry;
import java.util.regex.*;
import java.util.stream.*;

/**
 * Looks at the Jenkins job history for {@code job}, averages the time taken per test, printing any that took longer than
 * {@link QuickTest#EXPECTED_RUNTIME_THRESHOLD} - these are expected to be {@link SlowTest}s
 */
public class TestSlowTests {
    private static final HttpClient HTTP_CLIENT = HttpClient.newHttpClient();
    private static final Pattern TEST_PATTERN = Pattern
            .compile("\\[INFO\\]\\sTests\\srun:.+elapsed:\\s*(?<time>[\\d\\.]+).*in\\s+(?<test>.+)");

    private static String job = "Hazelcast-EE-pr-builder";
    private static String instance = "https://jenkins.hazelcast.com";
    private static String jenkinsCookie;

    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            throw new IllegalArgumentException("Requires Jenkins cookie as argument! Take the cookie from the browser...");
        } else {
            jenkinsCookie = args[0];

            Multimap<String, Double> testToTimes = getJobIDs().parallel().flatMap(id -> {
                try (Stream<String> stream = HTTP_CLIENT.send(getRequest(id + "/consoleText"), BodyHandlers.ofLines()).body()) {
                    return stream.collect(Collectors.toList()).stream();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }).map(TEST_PATTERN::matcher).filter(Matcher::matches).sequential()
                    .collect(ImmutableListMultimap.toImmutableListMultimap(matcher -> matcher.group("test"),
                            matcher -> Double.parseDouble(matcher.group("time"))));

            Map<String, Double> testToAverageTime = testToTimes.asMap().entrySet().stream().collect(Collectors
                    .toMap(Entry::getKey, entry -> entry.getValue().stream().mapToDouble(x -> x).average().orElseThrow()));

            String FORMAT = "|%-100s|%20s|%n";
            System.out.format(FORMAT, "Test Class", "Runtime (Seconds)");
            System.out.format(FORMAT, "-", "-");

            testToAverageTime.entrySet().stream()
                    .filter(entry -> entry.getValue() > QuickTest.EXPECTED_RUNTIME_THRESHOLD.getSeconds())
                    .sorted(Comparator.<Entry<String, Double>, Double> comparing(Entry::getValue).reversed()
                            .thenComparing(Entry::getKey))
                    .forEach(entry -> System.out.format(FORMAT, entry.getKey(),
                            NumberFormat.getInstance().format(entry.getValue())));
        }
    }

    private static Stream<String> getJobIDs() throws Exception {
        String response = HTTP_CLIENT.send(getRequest("api/json?tree=allBuilds[id]"), BodyHandlers.ofString()).body();

        try {
            return new ObjectMapper().readTree(response).get("allBuilds").findValues("id").stream().map(JsonNode::asText);
        } catch (JsonParseException e) {
            throw new RuntimeException(response, e);
        }
    }

    private static HttpRequest getRequest(String query) {
        return HttpRequest.newBuilder().header("Cookie", jenkinsCookie).uri(URI.create(instance + "/job/" + job + "/" + query))
                .build();
    }
}
```

Which lists the following tests run in `Hazelcast-pr-builder` above this threshold:
|Test Class                                                                                          |   Runtime (Seconds)|
|-                                                                                                   |                   -|
|com.hazelcast.sql.index.SqlIndexTest                                                                |             281.631|
|com.hazelcast.jet.kafka.impl.StreamKafkaPTest                                                       |             122.889|
|com.hazelcast.client.ClientRegressionWithRealNetworkTest                                            |             117.338|
|com.hazelcast.multimap.MultiMapSplitBrainTest                                                       |             111.444|
|com.hazelcast.jet.kafka.connect.KafkaConnectIntegrationTest                                         |             108.639|
|com.hazelcast.cardinality.CardinalityEstimatorSplitBrainTest                                        |             102.922|
|com.hazelcast.spi.impl.operationservice.impl.Invocation_TimeoutTest                                 |              97.868|
|com.hazelcast.internal.partition.GracefulShutdownTest                                               |              97.244|
|com.hazelcast.jet.mongodb.MongoSourceTest                                                           |              91.743|
|com.hazelcast.jet.kafka.impl.WriteKafkaPTest                                                        |              90.807|
|com.hazelcast.jet.kafka.impl.KafkaDataConnectionStressTest                                          |              85.778|
|com.hazelcast.internal.cluster.impl.AdvancedClusterStateTest                                        |              84.219|
|com.hazelcast.cp.internal.raft.impl.LocalRaftTest                                                   |                79.9|
|com.hazelcast.jet.sql.impl.connector.map.index.JetSqlIndexTest                                      |              79.825|
|com.hazelcast.spi.impl.operationservice.impl.RaftInvocationFailureTest                              |              75.094|
|com.hazelcast.cp.internal.CPGroupRebalanceTest                                                      |              74.869|
|com.hazelcast.client.cp.internal.datastructures.lock.FencedLockClientBasicTest                      |              74.032|
|com.hazelcast.internal.cluster.impl.MembershipFailureTest                                           |               71.04|
|com.hazelcast.client.cp.internal.datastructures.semaphore.SessionAwareSemaphoreClientBasicTest      |              70.959|
|com.hazelcast.cluster.LiteMemberJoinTest                                                            |              69.345|
|com.hazelcast.sql.SqlOrderByTest                                                                    |              68.977|
|com.hazelcast.cp.internal.datastructures.lock.FencedLockBasicTest                                   |              68.025|
|com.hazelcast.cp.internal.datastructures.semaphore.SessionAwareSemaphoreBasicTest                   |              67.459|
|classloading.ThreadLeakClientTest                                                                   |              67.387|
|com.hazelcast.client.cache.CacheClientListenerTest                                                  |              66.495|
|com.hazelcast.connector.Hz3SourcesTest                                                              |              63.871|
|com.hazelcast.client.cp.internal.datastructures.semaphore.SessionlessSemaphoreClientBasicTest       |               63.26|
|com.hazelcast.client.topic.ClientReliableTopicBlockTest                                             |              61.535|
|com.hazelcast.client.io.AdvancedNetworkClientIntegrationTest                                        |              60.435|
|com.hazelcast.spi.impl.operationservice.impl.Invocation_BlockingTest                                |               60.16|

**Changes:**
- Add Javadoc definition to `QuickTest`
- Made `HazelcastSerialClassRunner`/`HazelcastParallelClassRunner` log a warning to `System.err` if `QuickTest` goes above the threshold
   - `testMigrationStats_whenMigrationProcessCompletes is annotated as a QuickTest, expected to complete within 60 seconds - but took 120 seconds`
   - This isn't super obvious outside of Jenkins, and won't go onto the PR etc, but will flag when run locally at least
- Updated the `QuickTest`s to be `SlowTest`s.
   - Some already were, but the `pr-builder` happened to be building a backport to `4.1` when I collected this dataset

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/6654

### CODE REVIEW

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/index/SqlIndexTest.java
Comment (Fly-Style) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/index/SqlIndexTest.java:31]: This should be an exception, because it tests the _basic_ index scan logic.

**Review by Fly-Style — CHANGES_REQUESTED** [review]

**Comment by Fly-Style** [comment]
What we (SQL team) should definitely do - is somehow to enlight SQL index test.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/index/SqlIndexTest.java
Comment (JackPGreen) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/index/SqlIndexTest.java]: da66388a2fee9a9f64ab724e34f574aef33cd6b1

**Review by srknzl — APPROVED** [review]

**Review by Fly-Style — COMMENTED** [review]
Looks good to me.
I won't block the PR from merging, but let other teams members also approve changes for tests they are responsible.

File: hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java
Comment (frant-hartm) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: We have valid cases where we want to run a test taking more than 1 minute in PR builder. This will cause noise for such tests.

First I thought we can add an annotation parameter to `QuickTest` to mark such tests, but it is a class, not an annotation. So maybe introduce a new category or completely separate annotation which we can be used to mark long running tests which should run in PR builder and it would be checked here and we wouldn't print the log for such cases.

Wdyt? @vbekiaris @k-jamroz  your thoughts?

Comment (k-jamroz) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: maybe not report warning for `@Category({QuickTest.class, SlowTest.class})` (abusing `SlowTest` meaning)? 
BTW, didn't we consider removing `QuickTest.class` entirely, as it is (was before this PR) a de-facto no-op.

File: hazelcast/src/test/java/com/hazelcast/test/HazelcastSerialClassRunner.java
Comment (k-jamroz) [inline hazelcast/src/test/java/com/hazelcast/test/HazelcastSerialClassRunner.java:59]: note that code-coverage tests are significantly slower - we might get many warnings for them

File: hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java
Comment (k-jamroz) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: IMO it would be better to have an easy way to distinguish and find exceptional tests, so I'd prefer dedicated category or annotation

File: hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java
Comment (k-jamroz) [inline hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java:24]: "nightly builds and for code coverage" != !"PR builder and for code coverage"

Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java:24]: > "nightly builds and for code coverage" != !"PR builder and for code coverage"

Do you have a suggested alternative?
This isn't something I've changed?

Comment (k-jamroz) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java:24]: I referred to "opposite". They are not strictly opposite, third alternative being `NightlyTest`, so "The opposite of a `{@link QuickTest}`" might be misleading.

File: hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: > We have valid cases where we want to run a test taking more than 1 minute in PR builder. 

Did you have a specific test in mind? Just to check I haven't disabled it.

> This will cause noise for such tests

> a warning that nobody (?) will read

The primary case I was thinking of was where, while a feature is developed, the test starts as a `QuickTest` - but as more cases are identified, it gets slower and slower but no-one considers it's crossed the threshold. 

> First I thought we can add an annotation parameter to `QuickTest` to mark such tests, but it is a class, not an annotation. So maybe introduce a new category or completely separate annotation which we can be used to mark long running tests which should run in PR builder and it would be checked here and we wouldn't print the log for such cases.

That's a nice solution but I worry it's an extra layer of complexity - were probably better off not warning at all in that case.

> but if we really want to enforce the rule, we should fail the tests

The aim of this change is to make the PR builder nicer... I don't think adding more failures would make me very popular.


It doesn't sound like this check is well-received - I'm leaning towards removing it.

File: hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java
Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/SlowTest.java]: > I referred to "opposite". They are not strictly opposite, third alternative being `NightlyTest`, so "The opposite of a `{@link QuickTest}`" might be misleading.

Good spot, completely missed `NightlyTest`

180e8d852810ad387f5e5ff7ac45f94bb8106e31

File: hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java
Comment (k-jamroz) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: >> a warning that nobody (?) will read
> The primary case I was thinking of was where, while a feature is developed, the test starts as a QuickTest - but as more cases are identified, it gets slower and slower but no-one considers it's crossed the threshold.

Yes, I also see this phenomenon. During intensive development of some feature it is beneficial to have its tests running in the PR builder, but when the feature is mature and the development stops, and it is not a core feature that everything else depends on, the benefit of running all those tests in PR builder is much smaller, even if they take <1 minute each.

> > but if we really want to enforce the rule, we should fail the tests
> 
> The aim of this change is to make the PR builder nicer... I don't think adding more failures would make me very popular.

Yes, I agree that in current situation this would cause more harm than good.
 
> It doesn't sound like this check is well-received - I'm leaning towards removing it.

I am very in favour of making PR builder faster, and anything that leads to that is good. Of course I'd like to run all SQL and Jet tests in PR builds but this is infeasible.
BTW, I just had another idea: maybe each module should get a budget of time in PR builder (eg. 5min for SQL, 5 min for Jet, 10min for Core, etc) and each team could allocate it as they see fit: be it one big test or many small tests.

There is balance to find with moving tests from PR builds to nightly. We have currently imperfect process of handling failures in nightly test failures, so there is a compromise between speed of PR builds and speed and cost of fixing regressions. Another consideration is if we should run more tests for community contributions before merging, or which test might be important (mandatory?) for such PRs. This is much more complicated that just test runtime, but it is a first step.

Comment (frant-hartm) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: > Did you have a specific test in mind? Just to check I haven't disabled it.

E.g. Jet extensions tests (StreamKafkaPTest in your list) - each extension should be tested end to end for at least some happy path and common failure scenarios.

We should focus on tests in `hazelcast` module, which do not have `ParallelJVMTest` annotation - they take up their full time.

Other tests run in parallel in IIRC 8 threads so if you moving 80s test will give you only 10s on whole PR builder time.

The extension modules are also built in parallel, but there is a critical path, which used to be:

`hazelcast` <- `hazelcast-jet-kafka` <- `hazelcast-sql`

everything else outside of these modules doesn't matter, but maybe this changed - there is `hazelcast-jet-mongodb`, not sure if this one takes longer than kafka.

Comment (Fly-Style) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: > BTW, didn't we consider removing `QuickTest.class` entirely, as it is (was before this PR) a de-facto no-op.

I don't think it's a good idea. When you see a test category explicitly, at least you may think to re-classify the test if it consumes too much time. With implicit default, we will have more heavyweight tests added to the PR builder by mistake, due to the implicitness of test category.

Comment (JackPGreen) [reply hazelcast/src/test/java/com/hazelcast/test/annotation/QuickTest.java:40]: > Actually, failing your test if it doesn't have a defined category is a good idea. WDYT @JackPGreen?

I think it's good to have consistency - i.e. everywhere uses `QuickTest` or nowhere - but I think there's going to be a lot of consistency to iron out in the process.

**Pull request merged**
